opt: generate constrained partial index scans

mgartner opened the pull request:
This commit updates the GenerateConstrainedScans exploration rule so
that constrained partial index scans are now generated. For example,
consider the following index and query:

    CREATE INDEX ON t (a) WHERE b = 'foo'
    SELECT a FROM t WHERE b = 'foo' and a > 0 AND a < 10

GenerateConstrainedScans will generate the following query plan:

    project
     └── index-join t
          └── scan t@t_a_idx,partial
               └── constraint: /1/3: [/1 - /9]

The statistics builder and logical props builder were also updated to
correctly to correctly account for constrained partial index scans.

Fixes #50231 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52212)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 610 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDp6XP-8aCaYh-Xo5k4:-MDp6XP-8aCaYh-Xo5k5:b-brbia2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/statistics_builder.go#L610)):*
> ```Go
> 		// contain the query filter, partialidx.Implicator would have proven
> 		// non-implication, and a partial index scan would not have been
> 		// generated.
> ```

This makes sense in this example but I don't understand it in general.. Say we have `INDEX (x) WHERE (y >= 10)` and the index constraint is `x=1`. We want to estimate the number of rows in the table that have `x=1` AND `y >= 10`, not just `x=1`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 610 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDp6XP-8aCaYh-Xo5k4:-MDpAD4c-jCjhwG_2NNV:b-n5uxp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/statistics_builder.go#L610)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This makes sense in this example but I don't understand it in general.. Say we have `INDEX (x) WHERE (y >= 10)` and the index constraint is `x=1`. We want to estimate the number of rows in the table that have `x=1` AND `y >= 10`, not just `x=1`.
</blockquote></details>

You're right. I didn't make it clear that this only applies when the columns intersect—which the code is doing. I'll update the comment to make that more clear.

Also, I spent a bit of time thinking about if this holds true for all combinations of multi-column constraints. For example:

```
CREATE INDEX i ON t (x) WHERE x > 10 AND y > 100
SELECT * FROM t WHERE x > 50 AND y > 100
```

In this case, you want to apply 2 constraints to the scan: `x > 50` and `y > 100`. But this intersection rule still works fine, AFAIK, because the conjunction will be split into separate `FiltersItem`s in the predicate with separate constraints.

I've been unable to come up with an example where a multi-column constraint is possible here and needs special consideration.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 610 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDp6XP-8aCaYh-Xo5k4:-MDpEeay9azSBLExwnBC:bjqiyfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/statistics_builder.go#L610)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You're right. I didn't make it clear that this only applies when the columns intersect—which the code is doing. I'll update the comment to make that more clear.

Also, I spent a bit of time thinking about if this holds true for all combinations of multi-column constraints. For example:

```
CREATE INDEX i ON t (x) WHERE x > 10 AND y > 100
SELECT * FROM t WHERE x > 50 AND y > 100
```

In this case, you want to apply 2 constraints to the scan: `x > 50` and `y > 100`. But this intersection rule still works fine, AFAIK, because the conjunction will be split into separate `FiltersItem`s in the predicate with separate constraints.

I've been unable to come up with an example where a multi-column constraint is possible here and needs special consideration.
</blockquote></details>

Hm, intersection seems too "loose" of a check.. What about something like `INDEX (x,y) WHERE y IN (1,2,3)` and `SELECT WHERE x > 50 AND y IN (1,2,3)`? The index constraint is `[/51/1 - ]` but we still want to take into account that the index only contains a few values of `y`

Perhaps we need to check that all constraint spans have an "exact prefix" >= the index ordinal position of the constraint set's first column, or something to that effect


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice!

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MDqFRwxArnbVDNrE0Bj:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqBR1P1yfvUTYHI5G2:-MDqBR1P1yfvUTYHI5G3:b4k2i4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/statistics_builder.go#L612)):*
> ```Go
> 		// generated.
> 		if scan.Constraint != nil {
> 			nonConstraintFilters := make(FiltersExpr, 0, len(pred))
> ```

you could avoid this allocation if none of the predicate filters intersect with the scan constraint.
___
*[pkg/sql/opt/memo/statistics_builder.go, line 615 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqAmALExj6ggezamSW:-MDqAmALExj6ggezamSX:b-abvx03) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/statistics_builder.go#L615)):*
> ```Go
> 			for i := range pred {
> 				if constraintSet := pred[i].ScalarProps().Constraints; constraintSet != nil {
> 					if !constraintSet.ExtractCols().Intersects(scan.Constraint.Columns.ColSet()) {
> ```

I think you might want to use the `ConstrainedColumns` function instead of just using all of the columns from the scan's constraint. There might be columns (e.g., primary keys in a secondary index scan) that are part of the constraint but not actually constrained.
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqChmC49em0riv_0Ga:-MDqChmC49em0riv_0Gb:b-c6ef8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L157)):*
> ```
>       │    ├── columns: k:1(int!null) i:2(int!null) s:3(string)
>       │    ├── constraint: /2/3/1: [/216/'baz' - /229/'baz']
>       │    ├── stats: [rows=140, distinct(1)=140, null(1)=0, distinct(2)=14, null(2)=0, distinct(3)=47.0785632, null(3)=7.7]
> ```

looks like this row count is too high -- it's not taking the `s = 'baz'` predicate into account.
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqCwjF0iJrnC-ktuov:-MDqCwjF0iJrnC-ktuow:b-bvviq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L161)):*
> ```
>       │    └── fd: (1)-->(2,3)
>       └── filters
>            └── s:3 = 'baz' [type=bool, outer=(3), constraints=(/3: [/'baz' - /'baz']; tight), fd=()-->(3)]
> ```

this filter is unnecessary -- do we need to update the logic to determine tightness?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqChmC49em0riv_0Ga:-MDqoZq47k4RHZgvuHdT:b-zidqhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L157)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

looks like this row count is too high -- it's not taking the `s = 'baz'` predicate into account.
</blockquote></details>

I think this is fixed now.
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqCwjF0iJrnC-ktuov:-MDqoCmK3Mv4Vh7yN2IY:bfv721l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L161)):*
The scan constraint could allow rows that do not satisfy the constraint. For example:

> ```
> /2/3/1: [/216/'baz' - /229/'baz']
> ```

allows `i = 220, s = 'a'`. So isn't the filter necessary?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqChmC49em0riv_0Ga:-MDqwBAjDaljBQJl1zVy:b-1a5cou) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L157)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think this is fixed now.
</blockquote></details>

oops what I said was actually wrong.... it's not really constraining much with the `s = 'baz'` predicate since all the values in between i=216 and i=229 are still there. So I guess it was technically more correct before 😬
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqCwjF0iJrnC-ktuov:-MDqwjHz2q2TQkHN9FCO:b-tf7ck3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L161)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

The scan constraint could allow rows that do not satisfy the constraint. For example:

> ```
> /2/3/1: [/216/'baz' - /229/'baz']
> ```

allows `i = 220, s = 'a'`. So isn't the filter necessary?
</blockquote></details>

yep sorry -- my bad


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqChmC49em0riv_0Ga:-MDr092U3dhr5EEEyx-2:b-oxldmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L157)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

oops what I said was actually wrong.... it's not really constraining much with the `s = 'baz'` predicate since all the values in between i=216 and i=229 are still there. So I guess it was technically more correct before 😬
</blockquote></details>

But it should take into account the index predicate `s IN ('foo', 'bar', 'baz')`. Even though the scan is over `[/216/'baz' - /229/'baz']` its not over _all_ rows in that range, only over rows where s is `foo`, `bar`, or `baz`. Are you sure this isn't correct?


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/memo/testdata/stats/partial-index-scan, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MDqChmC49em0riv_0Ga:-MDrHGiH75vlAe7bXgF8:b-o8ek24) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ae1e9fde54da13c7f4e0e0f02fa51fff25589e9/pkg/sql/opt/memo/testdata/stats/partial-index-scan#L157)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

But it should take into account the index predicate `s IN ('foo', 'bar', 'baz')`. Even though the scan is over `[/216/'baz' - /229/'baz']` its not over _all_ rows in that range, only over rows where s is `foo`, `bar`, or `baz`. Are you sure this isn't correct?
</blockquote></details>

Ah! Sorry -- yep, that makes sense to me. `140 * 3/50 = 8.4`


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: generate constrained partial index scans

This commit updates the GenerateConstrainedScans exploration rule so
that constrained partial index scans are now generated. For example,
consider the following index and query:

    CREATE INDEX ON t (a) WHERE b = 'foo'
    SELECT a FROM t WHERE b = 'foo' and a > 0 AND a < 10

GenerateConstrainedScans will generate the following query plan:

    project
     └── index-join t
          └── scan t@t_a_idx,partial
               └── constraint: /1/3: [/1 - /9]

The statistics builder and logical props builder were also updated to
correctly to correctly account for constrained partial index scans.

Release note: None

mgartner commented on pkg/sql/opt/memo/logical_props_builder.go:
```diff
@@ -79,10 +85,15 @@ func (b *logicalPropsBuilder) buildScanProps(scan *ScanExpr, rel *props.Relation
 	if scan.Constraint != nil {
 		rel.NotNullCols.UnionWith(scan.Constraint.ExtractNotNullCols(b.evalCtx))
 	}
-	rel.NotNullCols.IntersectionWith(rel.OutputCols)
```
check_expr.go tests assertions are failing the tests because I've removed this line, causing the set of not-null cols to contain columns not in the RelExpr's output cols.

```
------- Stdout: -------
=== RUN   TestLogicalProps/lookup-join
--- FAIL: TestLogicalProps/lookup-join (0.00s)
opt_tester.go:454: testdata/logprops/lookup-join:24: not null cols (3) not a subset of output cols (1,2)
(1) assertion failure
Wraps: (2) attached stack trace
| github.com/cockroachdb/cockroach/pkg/sql/opt/props.(*Relational).Verify
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/props/verify.go:53
| github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).CheckExpr
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:38
| github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).MemoizeScan
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:16838
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructScan
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:249
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).pruneScanCols
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/prune_cols_funcs.go:331
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).PruneCols
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/prune_cols_funcs.go:249
| github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructProject
|   /go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1432
```

This is an interesting case because this appears to be the first time that columns can be guaranteed to be not null even though they aren't in the set out output cols. For example, if the partial index `INDEX (a) WHERE b > 0` is used for a scan, `b` is guaranteed to be non-null for any row returned by the scan, even though `b` is not contained in the index (hence not an output col).

If I add this line back in, it alters the stats for partial index scans such that they are slightly off. Notice in the example below the scan estimates ~25.1 rows while the index-join (originally a Select in that memo group) estimates ~24.8 rows. 

```
exec-ddl
CREATE INDEX idx ON a (i) WHERE s IN ('foo', 'bar', 'baz')
----

opt
SELECT * FROM a WHERE i > 10 AND i < 20 AND s IN ('foo', 'bar', 'baz')
----
index-join a
 ├── columns: k:1(int!null) i:2(int!null) s:3(string!null) t:4(string)
 ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── scan a@idx,partial
      ├── columns: k:1(int!null) i:2(int!null)
      ├── constraint: /2/1: [/11 - /19]
      ├── stats: [rows=25.1421429, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=25.1421429, distinct(2,3)=25.1421429, null(2,3)=0]
      ├── key: (1)
      └── fd: (1)-->(2)
```

Without this line, the stats seem more accurate:

```
exec-ddl
CREATE INDEX idx ON a (i) WHERE s IN ('foo', 'bar', 'baz')
----

opt
SELECT * FROM a WHERE i > 10 AND i < 20 AND s IN ('foo', 'bar', 'baz')
----
index-join a
 ├── columns: k:1(int!null) i:2(int!null) s:3(string!null) t:4(string)
 ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
 ├── key: (1)
 ├── fd: (1)-->(2-4)
 └── scan a@idx,partial
      ├── columns: k:1(int!null) i:2(int!null)
      ├── constraint: /2/1: [/11 - /19]
      ├── stats: [rows=24.8207143, distinct(2)=9, null(2)=0, distinct(3)=3, null(3)=0, distinct(2,3)=24.8207143, null(2,3)=0]
      ├── key: (1)
      └── fd: (1)-->(2)
```

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work! Thanks for all the time you put into this.

I think what you've done with the `NotNullCols` in logical props makes sense, but I've also identified the two places you could update in the stats code if you want to stick with the assumption that the logical property `NotNullCols` should be a subset of `OutputCols`. I'm hesitant to break that contract, so it may be better to add a bit of extra complexity to the stats code...

Maybe @andy-kimball or @raduberinde have an opinion about which is better.

> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF1Ae_gAVvVai-iva35:bwb2tyg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/memo/statistics_builder.go, line 720 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MF12q5864H2ut-FCfse:-MF12q5864H2ut-FCfsf:b-3kzcmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/335aef9ad0b8f0f2938e49ce1c60b410c1ba400e/pkg/sql/opt/memo/statistics_builder.go#L720)):*
> ```Go
> 	if constraint != nil {
> 		// Add any not-null columns from this constraint.
> 		notNullCols.UnionWith(constraint.ExtractNotNullCols(sb.evalCtx))
> ```

You could add notNullCols from the predicate here
___
*[pkg/sql/opt/memo/statistics_builder.go, line 2773 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-MF13KWI-4HKX5CsNdhV:-MF13KWI-4HKX5CsNdhW:bvygrpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/335aef9ad0b8f0f2938e49ce1c60b410c1ba400e/pkg/sql/opt/memo/statistics_builder.go#L2773)):*
> ```Go
> 	// Set null counts to 0 for non-nullable columns
> 	// ---------------------------------------------
> 	sb.updateNullCountsFromNotNullCols(relProps.NotNullCols, s)
> ```

instead of using relProps.NotNullCols, you could pass in notNullCols


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Well I was wrong... I tried your suggestions and removed the changes to the logical props builder, and there are still not-null columns that are not a subset of the output cols. I'm trying to figure out now what is causing the issue.

I don't feel too strongly, but it's a bit odd to consider scan constraints while building scan logical properties but not partial index predicates. They can provide similar guarantees. It seems like including not-null columns from the partial index predicate in logical properties would only lead to better plans because there is more information available to parent relational expressions.

I'm curious to know why this contract was created in the first place. Was it to catch minor bugs where not null columns were included that shouldn't be, or were there bad plans being created because there were not null columns that weren't in the output columns, and this was extra protection to prevent that from happening again?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF1OTkl4ocnW2ONLF0d:bl7wnq4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Having non-output columns in `NotNullCols` is a shaky concept IMO. How do you define NotNullCols precisely? Currently it's pretty simple: it's the subset of OutputCols for which this operator is known to not produce NULLs.

rytaft commented:
I didn't mean for you to remove all changes to the logical props builder -- I think all your changes are good except for the removal of the line to intersect with OutputCols.





<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> I didn't mean for you to remove all changes to the logical props builder -- I think all your changes are good except for the removal of the line to intersect with OutputCols.

Oops ya I got confused. Fixed.

> Having non-output columns in NotNullCols is a shaky concept IMO. How do you define NotNullCols precisely? Currently it's pretty simple: it's the subset of OutputCols for which this operator is known to not produce NULLs.

Even that definition isn't perfect. If you have `(IndexJoin (Scan t@partial))` then the IndexJoin _can be_ known to not produce NULLs if the partial index guarantees it. So: "it's the subset of OutputCols that are 1) not-null columns from the input or 2) columns for which this operator is known to filter out NULLs."

I'm fine with leaving this contract in place for now. But I am curious if propagating not-null non-output columns would help produce better plans in any cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF1UeSl5RHvC1YemcYW:bqeqh75)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I think the definition I wrote down above ("for which this operator is known to not produce NULLs") is correct. Your (1) point is not correct, e.g. outer joins can add NULLs.

> I am curious if propagating not-null non-output columns
In order to help answer that, we'd first need to know what "not-null non-output columns" are exactly :)

If it's something along the lines of "every row produced by this operator corresponds to a row in a table, and all those rows have (in the table) only non-NULL values on this other column from that table" then it is seems workable. Still, I'd probably put these in a separate field of the logical props.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF1cNiq2KVWN_ansKIz:b89c16n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Ahh good points. Thanks for the explanation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF1eA3RCMkR15pnzEzU:b-ooqdky)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52212#-:-MF2U7wzDiF1ThPvgkoB:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner commented:
bors r=rytaft
